test/gtest/ucs/test_memtype_cache.cc

```diff
@@ -47,38 +64,349 @@ class test_memtype_cache : public ucs::test_with_param<ucs_memory_type_t> {
+    test_ptr_released(ptr, size);
+    test_ptr_not_found(ptr, size);
+}
+
+UCS_TEST_P(test_memtype_cache, shared_page_regions) {
+    const std::vector<ucs_memory_type_t> supported_mem_types =
+        mem_buffer::supported_mem_types();
+    const size_t size = 1000000;
+    void *buf2_ptr, *cur_ptr;
+    std::vector<void*> initial_ptrs;
+
+    initial_ptrs.push_back(NULL);
+    initial_ptrs.push_back((void*)ucs_get_page_size());
+
+    for (std::vector<void*>::const_iterator initial_ptr = initial_ptrs.begin();
+         initial_ptr != initial_ptrs.end(); ++initial_ptr) {
+        for (std::vector<ucs_memory_type_t>::const_iterator iter =
+                 supported_mem_types.begin();
+             iter != supported_mem_types.end(); ++iter) {
+
+            std::vector<std::pair<void*, size_t> > released_ptrs;
+
+            /* Create two buffers that possibly will share one page
+             *
+             *                         < shared page >
+             *                            ||    ||
+             *                            \/    ||
+             *        +----------------------+  ||
+             * buf1:  |    |    |    |    |  |  \/
+             *        +----------------------+----------------------+
+             *                        buf2:  |  |    |    |    |    |
+             *                               +----------------------+
+             */
+            cur_ptr          = get_next_ptr(*initial_ptr, 0);
+            mem_buffer *buf1 = allocate_mem_buffer(cur_ptr, size, GetParam());
+
+            cur_ptr          = get_next_ptr(cur_ptr, size);
+            mem_buffer *buf2 = allocate_mem_buffer(size, *iter);
+
+            test_region_found(*buf1, size);
+            test_region_found(*buf2, size);
+
+            buf2_ptr = buf2->ptr();
+            release_mem_buffer(buf2, &released_ptrs);
+
+            test_ptr_released(buf2_ptr, size);
+
+            test_region_found(*buf1, size);
+
+            release_mem_buffer(buf1, &released_ptrs);
+
+            /* buffer `buf1` and `buf2` are released */
+            test_ptrs_released(&released_ptrs);
+        }
+    }
+}
+
+UCS_TEST_P(test_memtype_cache, different_mem_types) {
+    std::vector<void*> initial_ptrs;
+    std::vector<bool> insert_types;
+
+    initial_ptrs.push_back(NULL);
+    initial_ptrs.push_back((void*)ucs_get_page_size());
+
+    insert_types.push_back(false); /* push_back */
+    insert_types.push_back(true);  /* push_front */
+
+    const std::vector<ucs_memory_type_t> supported_mem_types =
+        mem_buffer::supported_mem_types();
+
+    /* The tests try to allocate/simulate two buffers with different memory types */
+    for (std::vector<void*>::const_iterator initial_ptr = initial_ptrs.begin();
+         initial_ptr != initial_ptrs.end(); ++initial_ptr) {
+        for (std::vector<ucs_memory_type_t>::const_iterator iter =
+                 supported_mem_types.begin();
+             iter != supported_mem_types.end(); ++iter) {
```

Comment:
can we just test HOST memory vs GetParam(), instead of all possible combinations?

Reply:
it doesn't take so long
why only `HOST` vs `GetParam()`? CUDA and CUDA_MANAGED has to be tested, isn't it?
or when RoCM will be supported fully in gtest, it would be great to test with RoCM memory type